GX-TimeFrequency/src/minimax_grids.F90

```diff
@@ -0,0 +1,498 @@
+  !! @param[inout] psi: corresponding auxiliary function (see transformation type definition)
+  !! @param[inout] mat_A: auxiliary matrix (see transformation type definition)  
+  !! @param[in] i_point: pointer for the current grid point
+  !! @param[inout] current_point:  current grid point ether omega(i_point) or tau_(i_point)
+  !! @param[in] transformation type :
+  !!        (1) the cosine transform cos(it) -> cos(iw): psi(omega,x), mat_A = cos(omega*tau)*psi(tau,x) 
+  !!        (2) the cosine transform cos(iw) -> cos(it): psi(tau,x)  , mat_A = cos(omega*tau)*psi(omega,x)
+  !!        (3) the sine transform   sin(it) -> sin(iw): psi(omega,x), mat_A = sin(omega*tau)*psi(tau,x)
+  subroutine calculate_psi_and_mat_A(num_points, tau_points, omega_points, num_x_nodes, x_mu, psi, &
+       mat_A, i_point, current_point, transformation_type)
+
+    integer, intent(in)                                :: num_points, num_x_nodes, i_point
+    real(kind=dp), allocatable, dimension(:), &
+         intent(in)                                    :: tau_points, omega_points, x_mu
+    real(kind=dp), allocatable, dimension(:), &          
+         intent(inout)                                 :: psi         
+    real(kind=dp), allocatable, dimension(:, :), &
+         intent(inout)                                 :: mat_A
+    real(kind=dp), intent(inout)                       :: current_point
+    integer, intent(in)                                :: transformation_type
+
+    ! Internal variables
+    integer                                            :: i_node, j_point
+    real(kind=dp)                                      :: tau, omega
+
+    ! Begin work
+
+    ! the cosine transform cos(it) -> cos(iw)
+    if (transformation_type.eq.1) then
+       omega = omega_points(i_point)
+       current_point = omega
+
+       ! psi(omega_k,x) = 2x/(x^2+omega_k^2)
+       do i_node = 1, num_x_nodes
+          psi(i_node) = 2.0_dp*x_mu(i_node)/((x_mu(i_node))**2 + omega**2)
+       end do
+
+       ! mat_A = cos(omega_k * tau) psi(tau,x)
+       do j_point = 1, num_points
+          tau = tau_points(j_point)
+          do i_node = 1, num_x_nodes
+             mat_A(i_node, j_point) = cos(omega*tau)*exp(-x_mu(i_node)*tau)
+          end do
+       end do
+
+    ! the cosine transform cos(iw) -> cos(it)  
+    else if (transformation_type.eq.2) then
+       tau = tau_points(i_point)
+       current_point = tau
+
+       ! psi(tau_k,x) = =exp(-x*|tau_k|)
+       do i_node = 1, num_x_nodes
+          psi(i_node) = exp(-x_mu(i_node)*tau)
+       end do
+
+       ! mat_A = cos(tau_k,omega) psi(omega,x)
+       do j_point = 1, num_points
+          omega = omega_points(j_point)
+          do i_node = 1, num_x_nodes
+             mat_A(i_node, j_point) = cos(tau*omega)*2.0_dp*x_mu(i_node)/(x_mu(i_node)**2 + omega**2)
+          end do
+       end do
+
+    ! the sine transform sin(it) -> sin(iw)         
+    else if (transformation_type.eq.3) then
```

Comment:
On a quick glance, the code for all of these if statements looks identical. You should just write it once as a subroutine, and pass the appropriate function in according what sort of transform it is.

i.e.

`cos(tau*omega)*2.0_dp*x_mu(i_node)/(x_mu(i_node)**2 + omega**2)`

`sin(omega*tau)*exp(-x_mu(i_node)*tau)`

etc

Reply:
Just in case you wanted to do this, rather than declare function interfaces in routine bodies (which I think looks ugly as hell), you can do:

 ```fortran
   ! Interface for function, allowing it to be used as a routine arg
    abstract interface 

        subroutine iterator_func(my_node)
            import Node
            class(Node), pointer, intent(inout) :: my_node
        end subroutine
    
     end interface

contains
     ! Routine example (just happens to be type-bound)
        subroutine iterate(this, iter)
            class(base_list_type), intent(inout) :: this
            !> User-supplied function to operate on list
            procedure(iterator_func) :: iter
    
            class(Node), pointer :: current_node 
            class(Node), pointer :: next_node
    
            current_node => this%first_node
            do while (associated(current_node))
                next_node => current_node%get_next()
                ! RELEVANT CALL
                call iter(current_node)
                current_node => next_node
            enddo
        end subroutine
```

Reply:
Hey @AlexBuccheri, I started to work of this already. It is not finished, but the OOP boilerplate is growing fast, see here 61fd94ef8534bc0e240a039e26ec6b846ad7f655

Reply:
Well, you cannot really call OOP what I am doing. I am just using pointers to procedures declared in an abstract interface. The thing is that it is getting verbose, because even when the three `if` cases are similar, there are several subtle differences. Do you have any idea how to make it more terse?

Reply:
What I'm suggesting isn't OOP, it's interface overloading (my example just happened to be a type-bound procedure, but that wasn't the point, the dependency injection was the point). As far as I can tell, the code is identical in the 3 cases except for the function that's being summed. So one writes that code once and passes the specific function in, according to the if-statement. 

I'll have to check it properly to confirm the code really is duplicated (i.e. I didn't notice _several_ differences). Will loo k at it properly tomorrow

Reply:
@AlexBuccheriIs it necessary to create another module? I can put the `tranformation_type` in minimax_utils 

@AlexBuccheri and @Panadestein , from the following analysis: 

[Transformations.pdf](https://github.com/nomad-coe/greenX/files/10472215/Transformations.pdf)

It seems to me that we only need three functions $\psi_{\mu}$, $\hat{\psi}_{\mu}$, and $\tilde{\psi}_{\mu}$. Also, keeping a copy of the matrix $A_{j \mu}$ we avoid the recalculations in the error evaluation. 

I just started with the implementation, I will let you know.

Reply:
Sure, you can put it in an existing module. I wouldn't put it in minimax_utils as they're not utility functions. I think that could be in minimax_grids.F90 (assuming no circular dependencies). But ultimately not too important

Reply:
I am going to mark this as resolved. I have prototyped a terse implementation of this subroutine that we could use at some point, if there is some interest. I remove the explicit loops and reformulate them in terms of outer procudts and Hadamard products:

```fortran
    ! the cosine transform cos(it) -> cos(iw)
    if (transformation_type == 1) then
       current_point = omega_points(i_point)
       psi = 2.0_dp * x_mu / (x_mu*x_mu + current_point*current_point)
       mat_A = cos(spread(current_point*tau_points, 1, num_x_nodes))&
             * exp(-outer_product(x_mu, tau_points))
    ! the cosine transform cos(iw) -> cos(it)  
    else if (transformation_type == 2) then
       current_point = tau_points(i_point)
       psi = exp(-x_mu*current_point)
       mat_A = 2.0_dp * cos(spread(current_point*omega_points, 1, num_x_nodes))&
             * lorentzian(spread(x_mu, 2, num_points),&
                          spread(omega_points, 1, num_x_nodes))
    ! the sine transform sin(it) -> sin(iw)         
    else if (transformation_type == 3) then
       current_point = omega_points(i_point)
       psi = 2.0_dp * current_point / (x_mu*x_mu + current_point*current_point)
       mat_A = sin(spread(current_point*tau_points, 1, num_x_nodes))&
             * exp(-outer_product(x_mu, tau_points))
    end if
```

with:

```fortran
  !> \brief Computes the outer product of two vectors
  !! @param[in] vac_a - first vector
  !! @param[in] vec_b - second vector
  function outer_product(vec_a, vec_b) result(prod)
    real(dp), dimension(:), intent(in)            :: vec_a, vec_b
    real(dp), dimension(size(vec_a), size(vec_b)) :: prod

    ! Local variables
    integer                                       :: len_a, len_b
    len_a = size(vec_a)
    len_b = size(vec_b)

    ! Hadamard product
    prod = spread(vec_a, 2, len_b) * spread(vec_b, 1, len_a)

  end function outer_product

  !> \brief Computes the value of the Lorentzian function
  !>        f(x, wt) = x^2 / (x^2 + y^2)
  !! @param[in] x - first argument
  !! @param[in] wt - second argument
  elemental function lorentzian (x, y) result(lor)
    real(dp)             :: lor
    real(dp), intent(in) :: x, y

    lor = x / (x*x + y*y)

  end function lorentzian
```
This code is already very terse, I think there is no need of simplifying this further. If there is interest, I can make a new PR afterward. I will merge the current one now.